mountinfo/mounted_linux_test.go

```diff
@@ -0,0 +1,254 @@
+	err = unix.Mount(src, mnt, "", unix.MS_BIND, "")
+	if err != nil {
+		err = &os.PathError{Op: "mount", Path: mnt, Err: err}
+		return
+	}
+	mounts = append(mounts, mnt)
+
+	// A regular file bind-mounted to itself.
+	mnt = filepath.Join(dir, "bind-mount-file")
+	err = ioutil.WriteFile(mnt, []byte(""), 0640)
+	if err != nil {
+		return
+	}
+
+	err = unix.Mount(mnt, mnt, "", unix.MS_BIND, "")
+	if err != nil {
+		err = &os.PathError{Op: "mount", Path: mnt, Err: err}
+		return
+	}
+	mounts = append(mounts, mnt)
+
+	// Not mounted socket.
+	sock := filepath.Join(dir, notMounted+".sock")
+	_, err = net.Listen("unix", sock)
+	if err != nil {
+		return
+	}
+	mounts = append(mounts, sock)
+
+	// Bind-mounted socket.
+	mnt = filepath.Join(dir, "bind-mounted-socket")
+	err = ioutil.WriteFile(mnt, []byte(""), 0640)
+	if err != nil {
+		return
+	}
+	err = unix.Mount(sock, mnt, "", unix.MS_BIND, "")
+	if err != nil {
+		err = &os.PathError{Op: "mount", Path: mnt, Err: err}
+		return
+	}
+	mounts = append(mounts, mnt)
+
+	// Not mounted directory.
+	mnt = filepath.Join(dir, notMounted+"-dir")
+	err = os.Mkdir(mnt, 0750)
+	if err != nil {
+		return
+	}
+	mounts = append(mounts, mnt)
+
+	// Not mounted file.
+	mnt = filepath.Join(dir, notMounted+"-file")
+	err = ioutil.WriteFile(mnt, []byte(""), 0640)
+	if err != nil {
+		return
+	}
+	mounts = append(mounts, mnt)
+
+	// A broken not-mounted symlink.
+	symlink := filepath.Join(dir, notMounted+"-broken-symlink")
+	err = unix.Symlink("non-existent-dest", symlink)
+	if err != nil {
+		err = &os.PathError{Op: "symlink", Path: symlink, Err: err}
+		return
+	}
+	mounts = append(mounts, symlink)
+
+	// A valid not-mounted symlink.
+	dst := filepath.Join(dir, "file")
+	err = ioutil.WriteFile(dst, []byte(""), 0640)
+	if err != nil {
+		return
+	}
+	symlink = filepath.Join(dir, notMounted+"-valid-symlink")
+	err = unix.Symlink(dst, symlink)
+	if err != nil {
+		err = &os.PathError{Op: "symlink", Path: symlink, Err: err}
+		return
+	}
+	mounts = append(mounts, symlink)
+
+	// A valid bind-mounted symlink
+	mnt = filepath.Join(dir, "bind-mounted-symlink")
+	err = ioutil.WriteFile(mnt, []byte(""), 0640)
+	if err != nil {
+		return
+	}
+	err = unix.Mount(symlink, mnt, "", unix.MS_BIND, "")
+	if err != nil {
+		err = &os.PathError{Op: "mount", Path: mnt, Err: err}
+		return
+	}
+	mounts = append(mounts, mnt)
+
+	return
+}
+
+func cleanupMounts(t *testing.T, dir string, mounts []string) {
+	for _, m := range mounts {
+		if strings.Contains(m, notMounted) {
+			continue
+		}
+		err := unix.Unmount(m, unix.MNT_DETACH)
+		if err != nil {
+			t.Logf("can't umount %s: %v", m, err)
+		}
+	}
+	if err := os.RemoveAll(dir); err != nil {
+		t.Log(err)
+	}
+}
+
+func TestMountedBy(t *testing.T) {
+	if os.Getuid() != 0 {
+		t.Skip("requires root")
+	}
+
+	dir, mounts, err := prepareMounts(t)
+	defer cleanupMounts(t, dir, mounts)
+	if err != nil {
+		t.Fatalf("setup failed: %v", err)
+	}
+
+	checked := false
+	for _, m := range mounts {
```

Comment:
(not a blocker, just thinking out loud as well) We could use subtests for these for easier debugging